src/Symfony/Component/String/AbstractString.php

```diff
@@ -0,0 +1,555 @@
+
+    /**
+     * @return static
+     */
+    abstract public function beforeLast(string $needle, bool $includeNeedle = false): self;
+
+    /**
+     * @return static[]
+     */
+    abstract public function chunk(int $length = 1): array;
+
+    /**
+     * @return static
+     */
+    public function collapseWhitespace(): self
+    {
+        $str = clone $this;
+        $str->string = trim(preg_replace('/(?:\s{2,}+|[^\S ])/', ' ', $str->string));
+
+        return $str;
+    }
+
+    abstract public function equalsTo(string $string): bool;
+
+    abstract public function endsWith(string $suffix): bool;
+
+    /**
+     * @return static
+     */
+    public function ensureLeft(string $prefix): self
+    {
+        return $this->startsWith($prefix) ? clone $this : $this->prepend($prefix);
+    }
+
+    /**
+     * @return static
+     */
+    public function ensureRight(string $suffix): self
+    {
+        return $this->endsWith($suffix) ? clone $this : $this->append($suffix);
+    }
+
+    public function isEmpty(): bool
+    {
+        return '' === $this->string;
+    }
+
+    public function isUtf8(): bool
+    {
+        return '' === $this->string || preg_match('//u', $this->string);
+    }
+
+    /**
+     * @return static
+     */
+    public function join(array $strings): self
+    {
+        $str = clone $this;
+
+        foreach ($strings as $k => $s) {
+            if ('' !== $s && !preg_match('//u', $s)) {
+                throw new InvalidArgumentException(sprintf('Invalid UTF-8 string at index "%s".', $k));
+            }
+        }
+
+        $str->string = implode($str->string, $strings);
+
+        return $str;
+    }
+
+    abstract public function length(): int;
+
+    public function match(string $pattern, int $flags = 0, int $offset = 0): ?array
+    {
+        if ('g' === $pattern[-1]) {
+            $pattern = substr($pattern, 0, -1);
+            $match = 'preg_match_all';
+        } else {
+            $match = 'preg_match';
+        }
+
+        if ($this->ignoreCase) {
+            $pattern .= 'i';
+        }
+
+        set_error_handler(static function ($t, $m) { throw new InvalidArgumentException($m); });
+
+        try {
+            if (false === $ret = $match($pattern.'u', $this->string, $matches, $flags, $offset)) {
+                $lastError = preg_last_error();
+
+                foreach (get_defined_constants(true)['pcre'] as $k => $v) {
+                    if ($lastError === $v && '_ERROR' === substr($k, -6)) {
+                        throw new RuntimeException('Matching failed with '.$k.'.');
+                    }
+                }
+
+                throw new RuntimeException('Matching failed with unknown error code.');
+            }
+        } finally {
+            restore_error_handler();
+        }
+
+        return $ret ? $matches : null;
+    }
+
+    abstract public function indexOf(string $needle, int $offset = 0): ?int;
```

Comment:
moving the thread about `contains` here:
> I suggest adding AbstractString::contains()

> why, when null !== $s->indexOf(...) provides the same?

> why include startsWith() if we can do 0 === $s->indexOf() ?

> startsWith can fail fast, while indexOf might traverse the whole string?

> $contains = $a->contains(...$strings); vs foreach($strings as $str) { if (null !== $a->indexOf($str)) { [...]

other APIs provide indexOfAny
we could make indexOf accept string|array for this purpose

Reply:
i believe you mean `indexOf(string|array $str): (?int)|array<?int>` ?
this way we will lose both parameter type hint and return type hint, i prefer we keep the type hint for `indexOf` :) 

i personally suggest adding : 

```php
public function contains(string ...$needles): bool
{
  $contains = false;
  foreach($needles as $needle) {
      if ('' === $needle || null !== $this->indexOf($needle)) {
        $contains = true;
        break;
      }
  }

  return $contains;
}
```

Reply:
I really mean `indexOf(string|array $needle): ?int`, meaning the first index of any of the strings

Reply:
> I suggest adding AbstractString::contains(...$needles)

this looses the $offset argument, which would make this method less useful

Reply:
i don't think losing `$offset` argument is an issue.
`$str->slice($offset)->contains(...$strings);`  is still better than going in a loop to the same :)

Reply:
`indexOf(string|string[] $needle): ?int` has my :+1:, it allows the same in a more performant way

Reply:
:-1: for sacrificing the type hint.

Reply:
`string|array` still won't enforce `string[]`. `string ...$needles` does.

Reply:
(actually, I'm going to leave that for a follow up PR)

Reply:
@nicholasruunu your variadic signature is dropping support for the offset feature.

Reply:
@stof Yeah, but seems like Nicolas was already ok with that since you can just `slice()` before.

Reply:
@nicolas-grekas explicitly said that he would accept the case with array *after* someone suggesting using `slice()` before.

Slicing the string first to extract a substring will have a performance impact compared to searching in the existing string from a non-0 offset (and it case your string is big, this can lead to a much bigger memory usage if you need to duplicate 99% of it into a substring).

Reply:
What @stof said: forcing a memory copy just to get an index is no-go on my side for a low-level API.

Reply:
Ah, I misunderstood. Yeah, I agree that the offset is useful, I digress. :)